File: Leetcode/1282. Group the People Given the Group Size They Belong To/solution.py
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        group_size_participants = {}

        for participant, group_size in enumerate(groupSizes):
            if group_size_participants.get(group_size):
                group_size_participants[group_size].append(participant)
            else:
                group_size_participants[group_size] = [participant]
        
        people_by_group_size = []

        for group_size, participants in group_size_participants.items():
            group_chunks = [participants[i:i+group_size] for i in range(0, len(participants), group_size)]
            people_by_group_size.extend(group_chunks)
    
        return people_by_group_size

File: Leetcode/1282. Group the People Given the Group Size They Belong To/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn group_the_people(group_sizes: Vec<i32>) -> Vec<Vec<i32>> {
        let mut group_size_participants: HashMap<i32, Vec<i32>> = HashMap::new();

        for (participant, group_size) in group_sizes.into_iter().enumerate()
        {
            group_size_participants
                .entry(group_size)
                .or_insert_with(Vec::new)
                .push(participant as i32)
        }

        let mut people_by_group_size: Vec<Vec<i32>> = Vec::new();

        for (group_size, participants) in group_size_participants
        {
            for chunk in participants.chunks(group_size as usize)
            {
                people_by_group_size.push(chunk.to_vec());
            }
        }

        people_by_group_size
    }
}